Ansible aws scale (#130)
microsoft/azure_arc: azure_arc_servers_jumpstart/docs/aws_scale_ansible.md

> # Onboard AWS servers to Azure Arc at scale using Ansible

Reviewer: ```suggestion
# Scaled Onboarding AWS EC2 instances to Azure Arc using Ansible
```

> The following README will guide you on how to automatically onboard servers to Azure Arc by using [Ansible](https://www.ansible.com/). This guide assumes that you have a basic understanding of Ansible. A basic Ansible playbook and configuration is provided that uses the [amazon.aws.aws_ec2](https://docs.ansible.com/ansible/latest/collections/amazon/aws/aws_ec2_inventory.html) plugin for dynamic loading of EC2 server inventory. 
> 

Reviewer: ```suggestion
The following README will guide you on how to automatically perform scaled onboarding of AWS EC2 instances to Azure Arc by using [Ansible](https://www.ansible.com/). 

This guide assumes that you have a basic understanding of Ansible. A basic Ansible playbook and configuration is provided that uses the [amazon.aws.aws_ec2](https://docs.ansible.com/ansible/latest/collections/amazon/aws/aws_ec2_inventory.html) plugin for dynamic loading of EC2 server inventory. 

```

> This guide can be used even if you do not already have an existing Ansible test environment. The guide includes a Terraform plan that will create a sample AWS EC2 server inventory comprised of four (4) Windows Server 2019 servers and four (4) Ubuntu servers along with a basic CentOS 7 Ansible control server with a simple configuration.

Reviewer: ```suggestion
This guide can be used even if you do not already have an existing Ansible test environment and includes a Terraform plan that will create a sample AWS EC2 server inventory comprised of four (4) Windows Server 2019 servers and four (4) Ubuntu servers along with a basic CentOS 7 Ansible control server with a simple configuration.
```

> ***Warning***: *The provided Ansible sample workbook uses WinRM with password authentication and HTTP to configure Windows-based servers. This is not advisable for production environments. If you are planning to use Ansible with Windows hosts in a production environment then you should use WinRM over HTTPS with a certificate.*

Reviewer: Maybe incl. a reference guide on how. Just a link...

# Prerequisites

* Clone this repo

    ```terminal
    git clone https://github.com/microsoft/azure_arc.git
    ```
    
* [Install or update Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). **Azure CLI should be running version 2.7** or later. Use ```az --version``` to check your current installed version.

* [Generate SSH Key](https://help.github.com/articles/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent/) (or use existing ssh key) 

* [Create free AWS account](https://aws.amazon.com/premiumsupport/knowledge-center/create-and-activate-aws-account/)

* [Install Terraform >=0.13](https://learn.hashicorp.com/terraform/getting-started/install.html)

## Create Azure Service Principal (SP)   

* To connect the AWS virtual machine to Azure Arc, an Azure Service Principal assigned with the "Contributor" role is required. To create it, login to your Azure account run the below command (this can also be done in [Azure Cloud Shell](https://shell.azure.com/)). 

>     ```bash

Reviewer: ```suggestion
    ```console
```

    az login
    az ad sp create-for-rbac -n "<Unique SP Name>" --role contributor
    ```

    For example:

>     ```az ad sp create-for-rbac -n "http://AzureArcAWS" --role contributor```

Reviewer: ```suggestion
```console
az ad sp create-for-rbac -n "http://AzureArcAWS" --role contributor
```
```

>     ```az ad sp create-for-rbac -n "http://AzureArcAWS" --role contributor```

Reviewer: ```suggestion
    ```console
    az ad sp create-for-rbac -n "http://AzureArcAWS" --role contributor
```
```

    Output should look like this:

>     ```
>     {
>     "appId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
>     "displayName": "AzureArcAWS",
>     "name": "http://AzureArcAWS",
>     "password": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
>     "tenant": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX"
>     }
>     ```

Reviewer: ```suggestion
    ```console
    {
    "appId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "displayName": "AzureArcAWS",
    "name": "http://AzureArcAWS",
    "password": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "tenant": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX"
    }
    ```
```

    **Note**: It is optional but highly recommended to scope the SP to a specific [Azure subscription and Resource Group](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest)

## Create an AWS identity

In order for Ansible to dynamically generate a server inventory from AWS, we will need to create a new AWS IAM role with appropriate permissions.

* Login to the [AWS management console](https://console.aws.amazon.com)

* After logging in, click the "Services" dropdown in the top left. Under "Security, Identity, and Compliance" select "IAM" to access the [Identity and Access Management page](https://console.aws.amazon.com/iam/home)

    ![](../img/aws_ubuntu/01.png) 

    ![](../img/aws_ubuntu/02.png)

* Click on "Users" from the left menu and then click on "Add user" to create a new IAM user.

    ![](../img/aws_ubuntu/03.png)

* On the "Add User" screen, give the user a name (e.g. "ansible") and select the "Programmatic Access" checkbox then click "Next"

    ![](../img/aws_ubuntu/04.png)

> * On the next "Set Permissions" screen, select "Attach existing policies directly" and then check the box next to AmazonEC2FullAccess as seen in the screenshot then click "Next"

Reviewer: ```suggestion
* On the next "Set Permissions" screen, select "Attach existing policies directly" and then check the box next to *AmazonEC2FullAccess* as seen in the screenshot then click "Next"
```

    ![](../img/aws_ubuntu/05.png)

* On the tags screen, assign a tag with a key of "azure-arc-demo" and click "Next" to proceed to the Review screen.

    ![](../img/aws_ubuntu/06.png)

* Double check that everything looks correct and click "Create user" when ready.

    ![](../img/aws_ubuntu/07.png)

* After the user is created, you will see the user's Access key ID and Secret access key. Copy these values down before clicking the Close button. In the screen below, you can see an example of what this should look like. Once you have these keys, you will be able to use them with Terraform to create AWS resources.

    ![](../img/aws_ubuntu/08.png)

# <a name="option1"></a>Option 1- Creating a sample AWS server inventory and Ansible control server using Terraform and onboarding the servers to Azure Arc

> *Note: If you already have an existing AWS server inventory and Ansible server, skip below to [Option 2](#option2).*

Reviewer: ```suggestion
**Note: If you already have an existing AWS server inventory and Ansible server, skip below to [Option 2](#option2).**
```

    ![](../img/aws_ubuntu/01.png) 

    ![](../img/aws_ubuntu/02.png)

* Click on "Users" from the left menu and then click on "Add user" to create a new IAM user.

    ![](../img/aws_ubuntu/03.png)

* On the "Add User" screen, give the user a name (e.g. "ansible") and select the "Programmatic Access" checkbox then click "Next"

    ![](../img/aws_ubuntu/04.png)

* On the next "Set Permissions" screen, select "Attach existing policies directly" and then check the box next to AmazonEC2FullAccess as seen in the screenshot then click "Next"

    ![](../img/aws_ubuntu/05.png)

* On the tags screen, assign a tag with a key of "azure-arc-demo" and click "Next" to proceed to the Review screen.

    ![](../img/aws_ubuntu/06.png)

* Double check that everything looks correct and click "Create user" when ready.

    ![](../img/aws_ubuntu/07.png)

* After the user is created, you will see the user's Access key ID and Secret access key. Copy these values down before clicking the Close button. In the screen below, you can see an example of what this should look like. Once you have these keys, you will be able to use them with Terraform to create AWS resources.

>     ![](../img/aws_ubuntu/08.png)

Reviewer: You might want to re-create your terraform user or mask out these secrets

Author: yes those were already rotated out.

    ```
    export AWS_ACCESS_KEY_ID="XXXXXXXXXXXXXXXXX"
    export AWS_SECRET_ACCESS_KEY="XXXXXXXXXXXXXXX"
    ```
* Run the Ansible playbook by executing the following command, substituting your Azure service principal id and service principal secret.

    ```
    ansible-playbook arc_agent.yml -i ansible_plugins/inventory_uswest2_aws_ec2.yml --extra-vars '{"service_principal_id": "XXXXXXX-XXXXX-XXXXXXX", "service_principal_secret": "XXXXXXXXXXXXXXXXXXXXXXXX"}'
    ```
    If the playbook run is successful, you should see output similar to the below screenshot. 

    ![](../img/aws_scale_ansible/05.png)

* Open Azure Portal and navigate to the Arc-AWS-Demo resource group. You should see the Azure Arc enabled servers listed.

    ![](../img/aws_scale_ansible/06.png)

## Clean up environment by deleting resources

To delete all the resources you created as part of this demo use the ```terraform destroy --auto-approve``` command as shown below.
    ![](../img/aws_scale_ansible/07.png)

# <a name="option2"></a>Option 2 - Onboarding an existing AWS server inventory to Azure Arc using your own Ansible control server

> *Note: If you do not have an existing AWS server inventory and Ansible server, navigate back to [Option 1](#option1).*

Reviewer: ```suggestion
**Note: If you do not have an existing AWS server inventory and Ansible server, navigate back to [Option 1](#option1).**
```

## Review provided Ansible configuration and playbook

> Navigate to the [ansible_config](../aws/scale_deployment/ansible/terraform/ansible_config) directory and review the provided configuration. The provided configuration contains a basic `ansible.cfg` file. This file enables the [amazon.aws.aws_ec2](https://docs.ansible.com/ansible/latest/collections/amazon/aws/aws_ec2_inventory.html) Ansible plugin which dynamically loads your server inventory by using an AWS IAM role. Ensure that the IAM role you are using has sufficient privileges to access the inventory you wish to onboard.

Reviewer: ```suggestion
Navigate to the [ansible_config](../aws/scale_deployment/ansible/terraform/ansible_config) directory and review the provided configuration. The provided configuration contains a basic *ansible.cfg* file. This file enables the [amazon.aws.aws_ec2](https://docs.ansible.com/ansible/latest/collections/amazon/aws/aws_ec2_inventory.html) Ansible plugin which dynamically loads your server inventory by using an AWS IAM role. Ensure that the IAM role you are using has sufficient privileges to access the inventory you wish to onboard.
```

The files in [group_vars](../aws/scale_deployment/ansible/terraform/ansible_config/group_vars) should be adjusted to provide the credentials you wish to use to onboard various ansible host groups.

When you have adjusted the provided config to support your environment, run the Ansible playbook by executing the following command, substituting your Azure service principal id and service principal secret.

```console
ansible-playbook arc_agent.yml -i ansible_plugins/inventory_uswest2_aws_ec2.yml --extra-vars '{"service_principal_id": "XXXXXXX-XXXXX-XXXXXXX", "service_principal_secret": "XXXXXXXXXXXXXXXXXXXXXXXX"}'
```
    
If the playbook run is successful, you should see output similar to the below screenshot. 
    ![](../img/aws_scale_ansible/05.png)

Open Azure Portal and navigate to the Arc-Aws-Demo resource group. You should see the Azure Arc enabled servers listed.

>     ![](../img/aws_scale_ansible/06.png)

Reviewer: Mask Subscription
